ext/random/random.c

```diff
@@ -0,0 +1,871 @@
+		}
+	}
+
+	return result % umax;
+}
+#endif
+
+zend_long php_random_range(php_random *random, zend_long min, zend_long max)
+{
+	zend_ulong umax = max - min;
+
+#if ZEND_ULONG_MAX > UINT32_MAX
+	if (umax > UINT32_MAX) {
+		return (zend_long) (range64(random, umax) + min);
+	}
+#endif
+
+	return (zend_long) (range32(random, umax) + min);
+}
+
+void php_random_array_data_shuffle(php_random *random, zval *array)
+{
+	uint32_t idx, j, n_elems, n_left;
+	Bucket *p, temp;
+	HashTable *hash;
+	zend_long rnd_idx;
+
+	n_elems = zend_hash_num_elements(Z_ARRVAL_P(array));
+
+	if (n_elems < 1) {
+		return;
+	}
+
+	hash = Z_ARRVAL_P(array);
+	n_left = n_elems;
+
+	if (EXPECTED(!HT_HAS_ITERATORS(hash))) {
+		if (hash->nNumUsed != hash->nNumOfElements) {
+			for (j = 0, idx = 0; idx < hash->nNumUsed; idx++) {
+				p = hash->arData + idx;
+				if (Z_TYPE(p->val) == IS_UNDEF) continue;
+				if (j != idx) {
+					hash->arData[j] = *p;
+				}
+				j++;
+			}
+		}
+		while (--n_left) {
+			rnd_idx = php_random_range(random, 0, n_left);
+			if (rnd_idx != n_left) {
+				temp = hash->arData[n_left];
+				hash->arData[n_left] = hash->arData[rnd_idx];
+				hash->arData[rnd_idx] = temp;
+			}
+		}
+	} else {
+		uint32_t iter_pos = zend_hash_iterators_lower_pos(hash, 0);
+
+		if (hash->nNumUsed != hash->nNumOfElements) {
+			for (j = 0, idx = 0; idx < hash->nNumUsed; idx++) {
+				p = hash->arData + idx;
+				if (Z_TYPE(p->val) == IS_UNDEF) continue;
+				if (j != idx) {
+					hash->arData[j] = *p;
+					if (idx == iter_pos) {
+						zend_hash_iterators_update(hash, idx, j);
+						iter_pos = zend_hash_iterators_lower_pos(hash, iter_pos + 1);
+					}
+				}
+				j++;
+			}
+		}
+		while (--n_left) {
+			rnd_idx = php_random_range(random, 0, n_left);
+			if (rnd_idx != n_left) {
+				temp = hash->arData[n_left];
+				hash->arData[n_left] = hash->arData[rnd_idx];
+				hash->arData[rnd_idx] = temp;
+				zend_hash_iterators_update(hash, (uint32_t)rnd_idx, n_left);
+			}
+		}
+	}
+	hash->nNumUsed = n_elems;
+	hash->nInternalPointer = 0;
+
+	for (j = 0; j < n_elems; j++) {
+		p = hash->arData + j;
+		if (p->key) {
+			zend_string_release_ex(p->key, 0);
+		}
+		p->h = j;
+		p->key = NULL;
+	}
+	hash->nNextFreeElement = n_elems;
+	if (!(HT_FLAGS(hash) & HASH_FLAG_PACKED)) {
+		zend_hash_to_packed(hash);
+	}
+}
+
+void php_random_string_shuffle(php_random *random, char *str, zend_long len)
+{
+	zend_long n_elems, rnd_idx, n_left;
+	char temp;
+	n_elems = len;
+
+	if (n_elems <= 1) {
+		return;
+	}
+
+	n_left = n_elems;
+
+	while (--n_left) {
+		rnd_idx = php_random_range(random, 0, n_left);
+		if (rnd_idx != n_left) {
+			temp = str[n_left];
+			str[n_left] = str[rnd_idx];
+			str[rnd_idx] = temp;
+		}
+	}
+}
+
+static inline zend_object *php_random_rng_common_init(zend_class_entry *ce, const php_random_rng_algo *algo, const zend_object_handlers *handlers) {
+	php_random_rng *rng = zend_object_alloc(sizeof(php_random_rng), ce);
```

Comment:
This seems to be correct - zend_object_alloc is zeroing out state (i.e. setting it to NULL)

```
/* Allocates object type and zeros it, but not the standard zend_object and properties.
 * Standard object MUST be initialized using zend_object_std_init().
 * Properties MUST be initialized using object_properties_init(). */
static zend_always_inline void *zend_object_alloc(size_t obj_size, zend_class_entry *ce) {
	void *obj = emalloc(obj_size + zend_object_properties_size(ce));
	memset(obj, 0, obj_size - sizeof(zend_object));
	return obj;
}
```

Reply:
Yes, but I'm still getting unintended Valgrind warnings.
Also, the ZTS build on Windows seems to be SEGVing during testing. I do not have a Windows environment, but will continue to try to resolve this.